platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/AbstractHashListener.java

```diff
@@ -97,80 +107,85 @@ protected AbstractHashListener(
+        synchronized (this) {
+            hashes.add(new VirtualHashRecord(path, hash));
+            if ((reconnectFlushInterval > 0) && (hashes.size() >= reconnectFlushInterval) && !flushInProgress.get()) {
+                dirtyHashesToFlush = hashes;
+                hashes = new ArrayList<>();
+                dirtyLeavesToFlush = leaves;
+                leaves = new ArrayList<>();
+            } else {
+                dirtyHashesToFlush = null;
+                dirtyLeavesToFlush = null;
+            }
+        }
+        if ((dirtyHashesToFlush != null) && (dirtyLeavesToFlush != null)) {
+            flush(dirtyHashesToFlush, dirtyLeavesToFlush);
+        }
     }
 
-    /**
-     * {@inheritDoc}
-     */
     @Override
-    public void onRankCompleted() {
-        batchLeaves.add(rankLeaves);
-        batchNodes.add(rankNodes);
+    public synchronized void onLeafHashed(final VirtualLeafRecord<K, V> leaf) {
+        leaves.add(leaf);
     }
 
-    /**
-     * {@inheritDoc}
-     */
     @Override
-    public void onBatchCompleted() {
-        long maxPath = -1;
-
-        Stream<VirtualHashRecord> sortedDirtyHashes = Stream.of();
-        for (int i = batchNodes.size() - 1; i >= 0; i--) {
-            final List<VirtualHashRecord> batch = batchNodes.get(i);
-            if (!batch.isEmpty()) {
-                sortedDirtyHashes = Stream.concat(sortedDirtyHashes, batch.stream());
-                maxPath = Math.max(maxPath, batch.get(batch.size() - 1).path());
-            }
+    public void onHashingCompleted() {
+        final List<VirtualHashRecord> finalNodesToFlush;
+        final List<VirtualLeafRecord<K, V>> finalLeavesToFlush;
+        synchronized (this) {
+            finalNodesToFlush = hashes;
+            hashes = null;
+            finalLeavesToFlush = leaves;
+            leaves = null;
         }
-
-        Stream<VirtualLeafRecord<K, V>> sortedDirtyLeaves = Stream.of();
-        for (int i = batchLeaves.size() - 1; i >= 0; i--) {
-            final List<VirtualLeafRecord<K, V>> batch = batchLeaves.get(i);
-            if (!batch.isEmpty()) {
-                sortedDirtyLeaves = Stream.concat(sortedDirtyLeaves, batch.stream());
-                maxPath = Math.max(maxPath, batch.get(batch.size() - 1).getPath());
-            }
+        if (!finalNodesToFlush.isEmpty() || !finalLeavesToFlush.isEmpty()) {
+            flush(finalNodesToFlush, finalLeavesToFlush);
         }
+    }
 
-        // flush it down
+    // Since flushes may take quite some time, this method is called outside synchronized blocks,
+    // otherwise all hashing tasks would be blocked on listener calls until flush is completed.
+    private void flush(
+            @NonNull final List<VirtualHashRecord> hashesToFlush,
+            @NonNull final List<VirtualLeafRecord<K, V>> leavesToFlush) {
+        if (!flushInProgress.compareAndSet(false, true)) {
+            throw new IllegalStateException("Cannot start flushing, flush already in progress?");
```

Comment:
Minor: `onNodeHashed` above checks the atomic flag w/o setting it. If the `onNodeHashed` is allowed to be called from multiple threads, then one of the threads might end up throwing an exception here even though a prior check in `onNodeHashed` succeeded because another thread might have started the flush in the meantime.

Reply:
This flag is not to protect from two (or more) threads handling `onNodeHashed()` in parallel. For this purpose, there is a synchronized block: only one of the threads may have `hashes.size() > reconnectFlushInterval`, while the other will have `hashes` empty. The flag is to handle a different case. Assume a flush is started (outside of the synchronized block), but it takes so long that there are reconnectFlushInterval (or more) hashes ready to flush again. The underlying MerkleDb is not ready to handle two parallel flushes yet, this is why the flag was introduced

Reply:
Yes, I understand that. My point is that that synchronized block calls the atomic.get() method only. However, the actual call to flush() happens outside of the block. So theoretically, there's a chance of calling flush() while another thread executing the onNodeHashed() would call it again because the flag hasn't been set yet.